=title Reboot using Ansible
=timestamp 2018-03-18T17:00:01
=indexes Ansible
=status show
=books ansible
=author szabgab
=archive 1
=comments_disqus_enable 0

=abstract start

Sometime we need to reboot the servers using Ansible

=abstract end

<h2>Ad-hoc command to reboot a single server</h2>

<code>
$ ansible -i inventory.cfg  192.168.56.11 -b -a "/sbin/shutdown -r now"
</code>

This reboots the server, but the reboot starts before Ansible finishes its work and thus it gives the following error message.

<code>
192.168.56.11 | UNREACHABLE! => {
    "changed": false,
    "msg": "Failed to connect to the host via ssh: Shared connection to 192.168.56.11 closed.\r\n",
    "unreachable": true
}
</code>


<h2>Ad-hoc command for delayed reboot</h2>

We can tell the <hl>shudown</hl> command to wait 1 minute (or more) and reboot only after that. (Probably 1-2 sec would be enought but as far as I know you cannot specify that for the <hl>shutdown command</hl>.) This gives enough time for Ansible to return a success report. It returns immediately and the machine shuts down (and because of the <hl>-r</hl> flag reboots) 1 minutes later.

<code>
$ ansible -i inventory.cfg  192.168.56.11 -b -a "/sbin/shutdown -r +1"
</code>

<code>
192.168.56.11 | SUCCESS | rc=0 >>
Shutdown scheduled for Sun 2018-03-18 17:01:40 IST, use 'shutdown -c' to cancel.
</code>

<h2>Reboot in a playbook</h2>

<code>
---
- hosts: all
  tasks:
     - command: /sbin/shutdown -r now
</code>

<code>
---
- hosts: all
  tasks:
     - name: reboot
      command: /sbin/shutdown -r now
</code>

<code>
ansible-playbook -i inventory.cfg --limit 192.168.56.11 -b reboot.yml
</code>

This was also discussed in the <a href="/getting-started-with-ansible-centos">Getting started with Ansible on CentOS</a> article.

<h2>Forgetting to become root</h2>

With all the explanation, the next time I tried to reboot a server I forgot to tell Ansible to <b>become root</b>.
Or in other words, I've forgotten to supply the <hl>-b</hl> flag.

<code>
$ ansible -i inventory.cfg 192.168.56.11 -a "/sbin/shutdown -r now"
</code>

The response was violent.

<code>
192.168.56.11 | FAILED | rc=1 >>
Failed to set wall message, ignoring: Interactive authentication required.
Failed to reboot system via logind: Interactive authentication required.
Failed to open /dev/initctl: Permission denied
Failed to talk to init daemon.non-zero return code
</code>

It took me a while to understand what happened, but afer a while I figure it out:

<code>
$ ansible -i inventory.cfg 192.168.56.11 -a "/sbin/shutdown -r now -b"
</code>

<h2>Comments</h2>


Gabor, you have a typo in last command.
the -b should be outside the " "

<hr>

Thank you for sharing :D This helped me with my jenkins job where I wanted to avoid the unreachable error because I was shutting down a machine.

<h2>

reboot starts before Ansible finishes its work ... see: https://stackoverflow.com/questions/29955605/how-to-reboot-centos-7-with-ansible
for async goodness. Also https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_async.html



